src/circuit/boolean.rs

```diff
@@ -375,35 +375,44 @@ impl Boolean {
         where E: Engine,
               CS: ConstraintSystem<E>
     {
-        let c = Self::xor(&mut cs, a, b)?;
+        match (a, b) {
+            (&Boolean::Constant(a), &Boolean::Constant(b)) => {
+                if a == b {
+                    Ok(())
+                } else {
+                    Err(SynthesisError::Unsatisfiable)
+                }
+            },
+            (&Boolean::Constant(true), a) | (a, &Boolean::Constant(true)) => {
+                cs.enforce(
+                    || "enforce equal to one",
+                    |lc| lc + CS::one(),
+                    |lc| lc + CS::one(),
+                    |_| a.lc(CS::one(), E::Fr::one())
```

Comment:
It's strictly better to do 0 * 0 = CS::one() - a.lc(), even though it doesn't help query densities.